[docs] SDK 50 guide for Sentry (#25712)
expo/expo: docs/pages/guides/using-sentry.mdx

export default Sentry.wrap(App);
```

</Step>

<Step label="6">

### Verify the configuration

Create a new release build of your app and verify that it uploads source maps correctly. You may want to add a button in your app to test that it is working and sourcemaps are wired up as expected, for example:

```jsx
import { Button } from 'react-native';

// Inside some component
<Button title="Press me" onPress={() => throw new Error('Hello, Sentry!')} />;
```

</Step>

## Usage with EAS Build

Ensure that `SENTRY_AUTH_TOKEN` is set in your build environment, and Sentry will automatically upload source maps for you. If you use environment variables rather than properties in your app config, ensure that those are set as well.

Using the above instructions, no additional work is needed to integrate Sentry into your project when using EAS Build.

## Usage with EAS Update

### Tag configuration

Configuring Sentry to tag your scope with information about your update allows to see errors happening on certain updates in Sentry dashboard.

Add the following snippet in the global scope as early as possible in your application's lifecycle.

```js
import * as Updates from 'expo-updates';

> const { metadata, extra } = Updates.manifest;

Author: cc @quinlanj

> const { metadata, extra } = Updates.manifest;

Reviewer: How about this: https://github.com/expo/expo/pull/25712#discussion_r1449709860

Open to suggestions for this suggestion 😄

> const { metadata, extra } = Updates.manifest;

Reviewer 2: Should be able to disambiguate the subtype by checking for the presence of `metadata`: https://github.com/expo/expo/blob/main/apps/expo-go/src/menu/DevMenuTaskInfo.tsx#L28

Something like:
```typescript
const { metadata, extra } =
    'metadata' in manifest ? manifest : { metadata: undefined, extra: undefined };
```
or this might be cleaner:
```typescript
const metadata = 'metadata' in manifest ? manifest.metadata : undefined;
const extra = 'metadata' in manifest ? manifest.extra : undefined;
```

`metadata` in the manifest has type `object` and `extra` has type `ManifestExtra`.

> const { metadata, extra } = Updates.manifest;

Reviewer: based on will's feedback: https://github.com/expo/expo/pull/25712#discussion_r1450825925

In this snippet, `updateGroup` is typed as unknown, so we do a runtime check for 'string' where necessary

> const { metadata, extra } = Updates.manifest;

Author: would this also preserve types?

```
const { manifest } = Updates;
const metadata = manifest?.metadata;
const extra = metadata ? manifest.extra : undefined;
const updateGroup = metadata?.updateGroup;
```

> const { metadata, extra } = Updates.manifest;

Reviewer 3: Can't we use, or cast to the exported `Manifest` type?
* https://github.com/expo/expo/blob/ab91257ace9c5a8196eacf6f7e2391f659538c15/packages/expo-updates/src/Updates.types.ts#L3

Or we can discern which exact type it is, by checking the presence of i.e. `metadata` key, via `in` operator before casting, if needed. So:

```tsx
const typedManifest = 'metadata' in manifest ? (manifest as ExpoUpdatesManifest) : (manifest as EmbeddedManifest); 
```

> const { metadata, extra } = Updates.manifest;

Author: this seems like a good suggestion @Simek. up to @quinlanj / @wschurman

> const { metadata, extra } = Updates.manifest;

Reviewer 2: It's also possible that it is empty object, though it's somewhat hard to tell when that's the case. Looks like it's when the library is disabled, either explicitly or due to an initialization error. So the actual type of `Updates.manifest` is `ExpoUpdatesManifest | EmbeddedManifest | {}` (I'll file a task to change it to not use `Partial<Manifest>`).

Sentry.init({
  // dsn, release, dist, etc...
});

Sentry.configureScope(scope => {
  scope.setTag('expo-update-group-id', metadata?.updateGroup);
  scope.setTag('expo-update-id', Updates.updateId);
  // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
  scope.setTag('expo-is-embedded-update', Updates.isEmbeddedLaunch);
  scope.setTag(
    'expo-update-debug-url',
    Updates.isEmbeddedLaunch
      ? 'not applicable for embedded updates'
      : `https://expo.dev/accounts/${extra?.expoClient?.owner}/projects/${extra?.expoClient?.slug}/updates/${metadata?.updateGroup}`
  );
});
```

Once configured, information about the associated update will show up in an error's tag section:

<ImageSpotlight alt="Sentry update tags" src="/static/images/guides/sentry-tags.png" />

### Upload source maps

After running `eas update`, upload the source maps to Sentry:

```sh

> SENTRY_AUTH_TOKEN=[super-secret-token] node node_modules/@sentry/react-native/scripts/expo-upload-sourcemaps.js dist

Reviewer 4: Let's use the `npx` command so we are not bound to the script path in the future. 

Plus it's shorter.

```suggestion
SENTRY_AUTH_TOKEN=[super-secret-token] npx sentry-expo-upload-sourcemaps dist
```

> ```

Reviewer 4: Would it make sense to move this above the `Tag configuration` section? I know it logically makes sense to first set tags, but I believe uploaded source maps are more important than the tags and I don't want users to overlook this.

Using the above instructions, no additional work is needed to integrate Sentry into your project when using EAS Build.

## Usage with EAS Update

After running `eas update`, upload the source maps to Sentry:

<Terminal
  cmd={[
    '# Pass in the `dist` directory generated by `eas update` to the upload script',
    '$ npx sentry-expo-upload-sourcemaps dist',
  ]}
/>

That's it! Errors for your updates will now be properly symbolicated in Sentry.

<Collapsible summary="Do you want to publish an update and the sourcemaps in one command?">

You can chain the commands together with `&&` to publish an update and upload the sourcemaps in one step:

<Terminal cmd={['$ eas update --channel <channel> && npx sentry-expo-upload-sourcemaps dist']} />

</Collapsible>

<Collapsible summary="Do you want to append additional update-related metadata to error reports?">

Configuring Sentry to tag your scope with information about your update allows to see errors happening on certain updates in Sentry dashboard.

Add the following snippet in the global scope as early as possible in your application's lifecycle.

> ```js
> import * as Updates from 'expo-updates';
> 
> const { metadata, extra } = Updates.manifest;

Reviewer: ```suggestion
```ts
import * as Updates from 'expo-updates';

// cast if using typescript
const { metadata, extra } = Updates.manifest as Record<string, any>;
```

> ```js
> import * as Updates from 'expo-updates';
> 
> const { metadata, extra } = Updates.manifest;
> 
> Sentry.init({
>   // dsn, release, dist, etc...
> });
> 
> Sentry.configureScope(scope => {
>   scope.setTag('expo-update-group-id', metadata?.updateGroup);
>   scope.setTag('expo-update-id', Updates.updateId);
>   // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
>   scope.setTag('expo-is-embedded-update', Updates.isEmbeddedLaunch);
>   scope.setTag(
>     'expo-update-debug-url',
>     Updates.isEmbeddedLaunch
>       ? 'not applicable for embedded updates'
>       : `https://expo.dev/accounts/${extra?.expoClient?.owner}/projects/${extra?.expoClient?.slug}/updates/${metadata?.updateGroup}`
>   );
> });

Reviewer: ```suggestion
```js
import * as Updates from 'expo-updates';

const manifest = Updates.manifest;
const metadata = "metadata" in manifest ? manifest.metadata : undefined;
const extra = "metadata" in manifest ? manifest.extra : undefined;
const updateGroup =
  metadata && "updateGroup" in metadata ? metadata.updateGroup : undefined;

Sentry.init({
  // dsn, release, dist, etc...
});

Sentry.configureScope((scope) => {
  if (typeof updateGroup === "string") {
    scope.setTag("expo-update-group-id", updateGroup);
  }
  scope.setTag("expo-update-id", Updates.updateId);
  // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
  scope.setTag("expo-is-embedded-update", Updates.isEmbeddedLaunch);
  scope.setTag(
    "expo-update-debug-url",
    Updates.isEmbeddedLaunch
      ? "not applicable for embedded updates"
      : `https://expo.dev/accounts/${extra?.expoClient?.owner}/projects/${extra?.expoClient?.slug}/updates/${updateGroup}`
  );
});
```

Author: > `${extra?.expoClient?.owner}`

will this possibly lead to `/accounts/undefined` ever? or is this owner field always actually present and this is just a typing quirk?

> `/updates/${updateGroup}`

it looks like `updateGroup` can be undefined, so this would lead to `/updates/undefined` if so

Reviewer 3: Can't we just use or cast to the exported `Manifest` type?
* https://github.com/expo/expo/blob/ab91257ace9c5a8196eacf6f7e2391f659538c15/packages/expo-updates/src/Updates.types.ts#L3

Reviewer: Amended suggestion: https://github.com/expo/expo/pull/25712#discussion_r1452783256

on the `owner` field -- the docs say that this should fall back to username but I tested this and it is indeed `undefined`. I've made a change where it falls back to `[account]`, and the website should handle this case. The edge case is if `slug` is undefined but `owner` isnt. But most manifests have `slug` so that field should be relatively safe.

> import * as Updates from 'expo-updates';
> 
> const { metadata, extra } = Updates.manifest;
> 
> Sentry.init({
>   // dsn, release, dist, etc...
> });
> 
> Sentry.configureScope(scope => {
>   scope.setTag('expo-update-group-id', metadata?.updateGroup);
>   scope.setTag('expo-update-id', Updates.updateId);
>   // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
>   scope.setTag('expo-is-embedded-update', Updates.isEmbeddedLaunch);
>   scope.setTag(
>     'expo-update-debug-url',
>     Updates.isEmbeddedLaunch
>       ? 'not applicable for embedded updates'
>       : `https://expo.dev/accounts/${extra?.expoClient?.owner}/projects/${extra?.expoClient?.slug}/updates/${metadata?.updateGroup}`
>   );
> });

Reviewer: ```suggestion
import * as Updates from 'expo-updates';

const { metadata, extra } = Updates.manifest;

Sentry.init({
  // dsn, release, dist, etc...
});

Sentry.configureScope(scope => {
  scope.setTag('expo-update-group-id', metadata?.updateGroup);
  scope.setTag('expo-update-id', Updates.updateId);
  // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
  scope.setTag('expo-is-embedded-update', Updates.isEmbeddedLaunch);
  scope.setTag(
    'expo-update-debug-url',
    Updates.isEmbeddedLaunch
      ? 'not applicable for embedded updates'
      : `https://expo.dev/accounts/${extra?.expoClient?.owner}/projects/${extra?.expoClient?.slug}/updates/${metadata?.updateGroup}`
  );
});
```
```suggestion
import * as Updates from 'expo-updates';

const manifest = Updates.manifest;
const metadata = "metadata" in manifest ? manifest.metadata : undefined;
const extra = "metadata" in manifest ? manifest.extra : undefined;
const updateGroup =
  metadata && "updateGroup" in metadata ? metadata.updateGroup : undefined;
  
Sentry.init({
  // dsn, release, dist, etc...
});

Sentry.configureScope((scope) => {
  if (typeof updateGroup === "string") {
    scope.setTag("expo-update-group-id", updateGroup);
  }
  scope.setTag("expo-update-id", Updates.updateId);
  // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
  scope.setTag("expo-is-embedded-update", Updates.isEmbeddedLaunch);
  if (Updates.isEmbeddedLaunch) {
    scope.setTag(
      "expo-update-debug-url",
      "not applicable for embedded updates"
    );
  } else if (typeof updateGroup === "string") {
    const owner = extra?.expoClient?.owner ?? "[account]";
    const slug = extra?.expoClient?.slug ?? "[project]";
    scope.setTag(
      "expo-update-debug-url",
      `https://expo.dev/accounts/${owner}/projects/${slug}/updates/${updateGroup}`
    );
  }
});
```

> import * as Updates from 'expo-updates';
> 
> const { metadata, extra } = Updates.manifest;
> 
> Sentry.init({
>   // dsn, release, dist, etc...
> });
> 
> Sentry.configureScope(scope => {
>   scope.setTag('expo-update-group-id', metadata?.updateGroup);
>   scope.setTag('expo-update-id', Updates.updateId);
>   // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
>   scope.setTag('expo-is-embedded-update', Updates.isEmbeddedLaunch);
>   scope.setTag(
>     'expo-update-debug-url',
>     Updates.isEmbeddedLaunch
>       ? 'not applicable for embedded updates'
>       : `https://expo.dev/accounts/${extra?.expoClient?.owner}/projects/${extra?.expoClient?.slug}/updates/${metadata?.updateGroup}`
>   );
> });

Reviewer: ```suggestion
import * as Updates from 'expo-updates';

const manifest = Updates.manifest;
const metadata = "metadata" in manifest ? manifest.metadata : undefined;
const extra = "metadata" in manifest ? manifest.extra : undefined;
const updateGroup =
  metadata && "updateGroup" in metadata ? metadata.updateGroup : undefined;

Sentry.init({
  // dsn, release, dist, etc...
});

Sentry.configureScope((scope) => {
  if (typeof updateGroup === "string") {
    scope.setTag("expo-update-group-id", updateGroup);
  }
  scope.setTag("expo-update-id", Updates.updateId);
  // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
  scope.setTag("expo-is-embedded-update", Updates.isEmbeddedLaunch);
  if (Updates.isEmbeddedLaunch) {
    scope.setTag(
      "expo-update-debug-url",
      "not applicable for embedded updates"
    );
  } else if (typeof updateGroup === "string") {
    const owner = extra?.expoClient?.owner ?? "[account]";
    const slug = extra?.expoClient?.slug ?? "[project]";
    scope.setTag(
      "expo-update-debug-url",
      `https://expo.dev/accounts/${owner}/projects/${slug}/updates/${updateGroup}`
    );
  }
});
```

Author: @quinlanj - what do you think about this for the `configureScope` section instead?

```js
Sentry.configureScope((scope) => {
  scope.setTag("expo-update-id", Updates.updateId);
  scope.setTag("expo-is-embedded-update", Updates.isEmbeddedLaunch);

  if (typeof updateGroup === "string") {
    scope.setTag("expo-update-group-id", updateGroup);

    const owner = extra?.expoClient?.owner ?? "[account]";
    const slug = extra?.expoClient?.slug ?? "[project]";
    scope.setTag(
      "expo-update-debug-url",
      `https://expo.dev/accounts/${owner}/projects/${slug}/updates/${updateGroup}`
    );
  } else if (Updates.isEmbeddedLaunch) {
    // This will be `true` if the update is the one embedded in the build, and not one downloaded from the updates server.
    if (Updates.isEmbeddedLaunch) {
      scope.setTag(
        "expo-update-debug-url",
        "not applicable for embedded updates"
      );
    }
  }
});
```

Reviewer: lgtm!

</Collapsible>

</Tab>

<Tab label="SDK 50 beta">

## Install and configure Sentry

<Step label="1">

### Sign up for a Sentry account and create a project

Before proceeding with installing Sentry, you'll need to make sure you have created a Sentry account and project. Here's how to do that:

<Step label="1.1">

[Sign up for Sentry](https://sentry.io/signup/) (it has a free tier), and create a project in your
Dashboard. Take note of your **organization slug**, **project name**, and **`DSN`** as you'll need
them later:

- **organization slug** is available in your **Organization settings** tab
- **project name** is available in your project's **Settings** > **Projects** tab (find it in the list)
- **`DSN`** is available in your project's **Settings** > **Projects** > **Project name** > **Client Keys
  (DSN)** tab.

</Step>

<Step label="1.2">

> Go to the [Sentry API section](https://sentry.io/settings/account/api/auth-tokens/), and create an
> **auth token**. The token requires the scopes: `org:read`, `project:releases`, and
> `project:write`. Save them.

Reviewer 4: Sentry has new organization auth tokens that are preferred for source maps uploaded as they are not bound to individuals but to org. 
```suggestion
Go to the [DEVELOPER SETTINGS -> Auth Tokens](https://sentry.io/settings/) and create a new token. The token is automatically scoped for Source Map Upload and Release Creation. Save it.
```

</Step>

Once you have each of these: organization slug, project name, DSN, and auth token, you're all set!

</Step>

<Step label="2">

### Installation

> In your project directory, run:

Reviewer 5: ```suggestion
Run the following command in your project directory to install `@sentry/react-native`:
```

Let's spell out what's happening in the command (good from accessibility view)

<Terminal cmd={['$ npx expo install @sentry/react-native']} />

</Step>

<Step label="3">

### App configuration

Configuring `@sentry/react-native` can be done through the config plugin in your [app config](/workflow/configuration/).  Add the following to your project's **app.json** or **app.config.js** file:

{/* prettier-ignore */}
```json app.json
{
  "expo": {

>     "plugins": ["sentry-expo", {

Reviewer 4: The new plugin import
```suggestion
    "plugins": ["@sentry/react-native/expo", {
```

> ```json app.json
> {
>   "expo": {
>     "plugins": ["sentry-expo", {
>       "organization": "sentry org slug, or use the `SENTRY_ORG` environment variable",
>       "project": "sentry project name, or use the `SENTRY_PROJECT` environment variable"
>     }]
>     /* @hide ... your existing configuration */ /* @end */
>   }
> }
> ```

Reviewer: Should be nested array as per https://docs.expo.dev/config-plugins/introduction/#use-a-config-plugin
```suggestion
```json app.json
{
  "expo": {
    "plugins": [["@sentry/react-native", {
      "organization": "sentry org slug, or use the `SENTRY_ORG` environment variable",
      "project": "sentry project name, or use the `SENTRY_PROJECT` environment variable"
    }]
    /* @hide ... your existing configuration */ /* @end */
    ]
  }
}
```
```

Next, in any environment where you would like to be able to create releases and upload sourcemaps to Sentry, you will need to set the `SENTRY_AUTH_TOKEN` environment variable to your [Sentry auth token](https://docs.sentry.io/product/cli/configuration/). [Learn how to create a token](https://sentry.io/settings/account/api/).

If you are using EAS Build, you can set the `SENTRY_AUTH_TOKEN` environment variable by [creating a secret of the same name](/build-reference/variables/#using-secrets-in-environment-variables).

> > **warning** Never commit secrets to your repository. Keep the Sentry auth token in a safe place, outside your repository.

Reviewer 5: I feel, we should provide more context on what that safe place can be, may be can mention one or two examples here?

<ConfigReactNative>

> If you do not use [Continuous Native Generation (CNG)](https://docs.expo.dev/workflow/continuous-native-generation/), then you should use the [@sentry/wizard](https://docs.sentry.io/platforms/react-native/#install).

Reviewer 5: ```suggestion
If you do not use [Continuous Native Generation (CNG)](https://docs.expo.dev/workflow/continuous-native-generation/), then you should use the [`@sentry/wizard`](https://docs.sentry.io/platforms/react-native/#install).
```

Since this is a library, let's use inline syntax highlighting here.

> </ConfigReactNative>

Reviewer 4: This is true, but users can also use it to set up apps with CNG.

https://github.com/getsentry/sentry-wizard/pull/505

</Step>

<Step label="4">

### Initialize Sentry

Add the following to your app's main file such as **App.js**:

```js
import * as Sentry from '@sentry/react-native';

Sentry.init({
  dsn: 'YOUR DSN HERE',
  debug: true, // If `true`, Sentry will try to print out useful debugging information if something goes wrong with sending the event. Set it to `false` in production
});
```

Now wrap the root component of your app with Sentry. This may be **App.js** or **index.js** depending on your project, or in [**app/_layout.js** with Expo Router](https://docs.expo.dev/router/layouts/#create-a-layout-route).

```js
import * as Sentry from "@sentry/react-native";

> // your App component here

Reviewer 5: ```suggestion
// Your App component here
```

We start all comments with captialziation.

export default Sentry.wrap(App);
```

</Step>

<Step label="5">

### Verify the configuration

Create a new release build of your app and verify that it uploads source maps correctly. You may want to add a button in your app to test that it is working and sourcemaps are wired up as expected, for example:

> ```js

Reviewer 3: ```suggestion
```jsx
```

Nit, for betted code highlight. Also setting the code block title might be nice, for clarity, and to allow easier copy.

import { Button } from 'react-native';

> // inside some component

Reviewer 5: ```suggestion
// Inside some component
```

<Button title='Press me' onPress={() => throw new Error('Hello, Sentry!')} />
```

</Step>

## Usage with EAS Build

Using the above instructions, no additional work is needed to integrate Sentry into your project when using EAS Build.

## Usage with EAS Update

> In the current state of SDK 50 beta, EAS Update is not yet supported. We're working on making it easier than ever to integrate updates with Sentry, and we plan to release along with SDK 50 proper.

Reviewer 5: ```suggestion
In the current state of SDK 50 beta, EAS Update is not yet supported. We're working on making it easier than ever to integrate updates with Sentry, and we plan to release it along with SDK 50.
```
